### PR TITLE
Fix incorrect event ordering

### LINKED ISSUE #985: bug: Events not emited in correct order when call to other contracts
By reading the code, I'm pretty sure we don't emit the event in the correct order when a contract call another one.

## Problem
Here is an example of flow:

tx call contract1
 - event 1 is emitted
 -contract1 call contract2:
     -event 2 is emitted
 - event 3 is emitted

 What we expect:
 the event should be emitted in the following order: 1, 2, 3
 
 What is happening:
 event are emitted in this order: 1, 3, 2
 
 ## What to do:
 
 1) Write test to prove my point. Related to: #648
 2) Fix it
 
 ### How to fix
 
 ```rust
pub struct OrderedEvent {
    pub order: usize,
    pub event: EventContent,
}
```
 
 Events are emitted after tx execution by reading the returned `CallInfo` struct. It contains a `CallExecution` struct that contains a `Vec<OrderedEvent>`. All the events in this vec are guaranteed to be in ascending order. But they are not necessarily contiguous. Going back to our example it will look like this `[event1, event3]`, while `event2` is stored in another `CallInfo` in the field `inner_calls`.
 
 Rather than first emitting all current `CallInfo` events and then recursively iterating over the `inner_calls`, we should read all of them at the same time and order them.
 
 Two possible implems:
 1) visit all `CallInfo` and store their events in a single vec and then order by their field `order`. The drawback is allocation/reallocation, vec sorting and another extra iteration to then emit the events.
 2) visit all `CallInfo`, keep a trace of their vec of events somewhere, and use those pointers to emit them in the correct order. Advantages: no allocation.

### PR DESCRIPTION
**Pull request by m-kus**
Resolves #985 

This patch changes the way `CallInfo` structure is traversed when contract events are aggregated. 

Before:
- Emit current call events 
- Then recursively handle all inner calls

After:
- Emit current call events while their orders are sequential (increase by one)
- Otherwise try to find the continuation of the sequence in the next inner call
- Repeat again until all events & inner calls are handled

Notes:
- We trust blockifier that in every external/inner call events will be sorted and form a correct sequence

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,7 @@
 - feat: add a `genesis_loader` for the node and mocking
 - feat: add `madara_tsukuyomi` as a submodule
 - branding: use new logo in the README
+- fix: events are emitted in correct sequential order
```
Comment (tdelabro) [inline CHANGELOG.md:21]: You added stuffs to the README too :)

File: cairo-contracts/src/test_contracts/emit_multiple_events_across_contracts.cairo [hunk 2]
```diff
@@ -39,3 +39,14 @@ func emit_external{ syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_chec
     IExternalContract.emit_external(contract_address=_external_contract_addr);
     return();
 }
+
+@external
+func emit_sandwich{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
```
Comment (tdelabro) [inline cairo-contracts/src/test_contracts/emit_multiple_events_across_contracts.cairo:44]: 🥪

File: crates/pallets/starknet/src/lib.rs [hunk 3]
```diff
@@ -1066,28 +1066,61 @@ impl<T: Config> Pallet<T> {
...
+    /// * `next_order` — Next expected event order, has to be 0 for a top level invocation
     ///
     /// # Returns
     ///
-    /// The result of the operation.
+    /// Next expected event order
     #[inline(always)]
-    fn emit_events(call_info: &mut CallInfo) -> Result<Vec<StarknetEventType>, EventError> {
-        let mut events = Vec::new();
+    fn emit_events(
```
Comment (tdelabro) [inline crates/pallets/starknet/src/lib.rs:1077]: Maybe change name to `emit_events_in_call_info` for more precision

File: cairo-contracts/src/test_contracts/emit_multiple_events_across_contracts.cairo [hunk 4]
```diff
@@ -39,3 +39,14 @@ func emit_external{ syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_chec
...
+    let (_external_contract_addr) = external_contract_addr.read();
+    internal.emit();
+    IExternalContract.emit_external(contract_address=_external_contract_addr);
+    internal.emit();
+    IExternalContract.emit_external(contract_address=_external_contract_addr);
```
Comment (tdelabro) [inline cairo-contracts/src/test_contracts/emit_multiple_events_across_contracts.cairo:49]: Just to be 100% sure, can you confirm that those two calls will result in two different `CallInfo` inside the `inner_calls` fields? Even if those are calls to the same contract

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: cairo-contracts/src/test_contracts/emit_multiple_events_across_contracts.cairo (hunk 4 above)
Comment (m-kus) [reply cairo-contracts/src/test_contracts/emit_multiple_events_across_contracts.cairo:49]: Yep, here is the execution call info:

```
CallInfo {
    call: CallEntryPoint {
        class_hash: Some(
            ClassHash(
                StarkFelt(
                    "0x0279d77db761fba82e0054125a6fdb5f6baa6286fa3fb73450cc44d193c2d37f",
                ),
            ),
        ),
        code_address: None,
        entry_point_type: External,
        entry_point_selector: EntryPointSelector(
            StarkFelt(
                "0x015d40a3d6ca2ac30f4031e42be28da9b056fef9bb7357ac5e85627ee876e5ad",
            ),
        ),
        calldata: Calldata(
            [
                StarkFelt(
                    "0x051a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02cf",
                ),
                StarkFelt(
                    "0x00d44e726c9388338336a865ac2ad319196e507965cce7fee90b116837b795f7",
                ),
                StarkFelt(
                    "0x0000000000000000000000000000000000000000000000000000000000000000",
                ),
            ],
        ),
        storage_address: ContractAddress(
            PatriciaKey(
                StarkFelt(
                    "0x01a3339ec92ac1061e3e0f8e704106286c642eaf302e94a582e5f95ef5e6b4d0",
                ),
            ),
        ),
        caller_address: ContractAddress(
            PatriciaKey(
                StarkFelt(
                    "0x0000000000000000000000000000000000000000000000000000000000000000",
                ),
            ),
        ),
        call_type: Call,
        initial_gas: 10000000000,
    },
    execution: CallExecution {
        retdata: Retdata(
            [],
        ),
        events: [],
        l2_to_l1_messages: [],
        failed: false,
        gas_consumed: StarkFelt(
            "0x0000000000000000000000000000000000000000000000000000000000000000",
        ),
    },
    vm_resources: ExecutionResources {
        n_steps: 310,
        n_memory_holes: 0,
        builtin_instance_counter: {
            "range_check_builtin": 1,
        },
    },
    inner_calls: [
        CallInfo {
            call: CallEntryPoint {
                class_hash: Some(
                    ClassHash(
                        StarkFelt(
                            "0x0000000000000000000000000000000000000000000000000000000000092000",
                        ),
                    ),
                ),
                code_address: Some(
                    ContractAddress(
                        PatriciaKey(
                            StarkFelt(
                                "0x051a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02cf",
                            ),
                        ),
                    ),
                ),
                entry_point_type: External,
                entry_point_selector: EntryPointSelector(
                    StarkFelt(
                        "0x00d44e726c9388338336a865ac2ad319196e507965cce7fee90b116837b795f7",
                    ),
                ),
                calldata: Calldata(
                    [],
                ),
                storage_address: ContractAddress(
                    PatriciaKey(
                        StarkFelt(
                            "0x051a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02cf",
                        ),
                    ),
                ),
                caller_address: ContractAddress(
                    PatriciaKey(
                        StarkFelt(
                            "0x01a3339ec92ac1061e3e0f8e704106286c642eaf302e94a582e5f95ef5e6b4d0",
                        ),
                    ),
                ),
                call_type: Call,
                initial_gas: 10000000000,
            },
            execution: CallExecution {
                retdata: Retdata(
                    [],
                ),
                events: [
                    OrderedEvent {
                        order: 0,
                        event: EventContent {
                            keys: [
                                EventKey(
                                    StarkFelt(
                                        "0x02c99ace14ee0a5ef932dc609df0943ab7ac16b7583634612f8dc35a4289a6ce",
                                    ),
                                ),
                            ],
                            data: EventData(
                                [],
                            ),
                        },
                    },
                    OrderedEvent {
                        order: 2,
                        event: EventContent {
                            keys: [
                                EventKey(
                                    StarkFelt(
                                        "0x02c99ace14ee0a5ef932dc609df0943ab7ac16b7583634612f8dc35a4289a6ce",
                                    ),
                                ),
                            ],
                            data: EventData(
                                [],
                            ),
                        },
                    },
                    OrderedEvent {
                        order: 4,
                        event: EventContent {
                            keys: [
                                EventKey(
                                    StarkFelt(
                                        "0x02c99ace14ee0a5ef932dc609df0943ab7ac16b7583634612f8dc35a4289a6ce",
                                    ),
                                ),
                            ],
                            data: EventData(
                                [],
                            ),
                        },
                    },
                ],
                l2_to_l1_messages: [],
                failed: false,
                gas_consumed: StarkFelt(
                    "0x0000000000000000000000000000000000000000000000000000000000000000",
                ),
            },
            vm_resources: ExecutionResources {
                n_steps: 271,
                n_memory_holes: 0,
                builtin_instance_counter: {},
            },
            inner_calls: [
                CallInfo {
                    call: CallEntryPoint {
                        class_hash: Some(
                            ClassHash(
                                StarkFelt(
                                    "0x0000000000000000000000000000000000000000000000000000000000091000",
                                ),
                            ),
                        ),
                        code_address: Some(
                            ContractAddress(
                                PatriciaKey(
                                    StarkFelt(
                                        "0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02cf",
                                    ),
                                ),
                            ),
                        ),
                        entry_point_type: External,
                        entry_point_selector: EntryPointSelector(
                            StarkFelt(
                                "0x000a5a5773e9e2a4cae1ba2eb67c56b6f5eaa53170cea3df7e71c057f049e957",
                            ),
                        ),
                        calldata: Calldata(
                            [],
                        ),
                        storage_address: ContractAddress(
                            PatriciaKey(
                                StarkFelt(
                                    "0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02cf",
                                ),
                            ),
                        ),
                        caller_address: ContractAddress(
                            PatriciaKey(
                                StarkFelt(
                                    "0x051a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02cf",
                                ),
                            ),
                        ),
                        call_type: Call,
                        initial_gas: 10000000000,
                    },
                    execution: CallExecution {
                        retdata: Retdata(
                            [],
                        ),
                        events: [
                            OrderedEvent {
                                order: 1,
                                event: EventContent {
                                    keys: [
                                        EventKey(
                                            StarkFelt(
                                                "0x03ee3b73373acd583a130924aad6dc38cfdc44ba0555ba94ce2ff63980ea0632",
                                            ),
                                        ),
                                    ],
                                    data: EventData(
                                        [],
                                    ),
                                },
                            },
                        ],
                        l2_to_l1_messages: [],
                        failed: false,
                        gas_consumed: StarkFelt(
                            "0x0000000000000000000000000000000000000000000000000000000000000000",
                        ),
                    },
                    vm_resources: ExecutionResources {
                        n_steps: 46,
                        n_memory_holes: 0,
                        builtin_instance_counter: {},
                    },
                    inner_calls: [],
                    storage_read_values: [],
                    accessed_storage_keys: {},
                },
                CallInfo {
                    call: CallEntryPoint {
                        class_hash: Some(
                            ClassHash(
                                StarkFelt(
                                    "0x0000000000000000000000000000000000000000000000000000000000091000",
                                ),
                            ),
                        ),
                        code_address: Some(
                            ContractAddress(
                                PatriciaKey(
                                    StarkFelt(
                                        "0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02cf",
                                    ),
                                ),
                            ),
                        ),
                        entry_point_type: External,
                        entry_point_selector: EntryPointSelector(
                            StarkFelt(
                                "0x000a5a5773e9e2a4cae1ba2eb67c56b6f5eaa53170cea3df7e71c057f049e957",
                            ),
                        ),
                        calldata: Calldata(
                            [],
                        ),
                        storage_address: ContractAddress(
                            PatriciaKey(
                                StarkFelt(
                                    "0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02cf",
                                ),
                            ),
                        ),
                        caller_address: ContractAddress(
                            PatriciaKey(
                                StarkFelt(
                                    "0x051a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02cf",
                                ),
                            ),
                        ),
                        call_type: Call,
                        initial_gas: 10000000000,
                    },
                    execution: CallExecution {
                        retdata: Retdata(
                            [],
                        ),
                        events: [
                            OrderedEvent {
                                order: 3,
                                event: EventContent {
                                    keys: [
                                        EventKey(
                                            StarkFelt(
                                                "0x03ee3b73373acd583a130924aad6dc38cfdc44ba0555ba94ce2ff63980ea0632",
                                            ),
                                        ),
                                    ],
                                    data: EventData(
                                        [],
                                    ),
                                },
                            },
                        ],
                        l2_to_l1_messages: [],
                        failed: false,
                        gas_consumed: StarkFelt(
                            "0x0000000000000000000000000000000000000000000000000000000000000000",
                        ),
                    },
                    vm_resources: ExecutionResources {
                        n_steps: 46,
                        n_memory_holes: 0,
                        builtin_instance_counter: {},
                    },
                    inner_calls: [],
                    storage_read_values: [],
                    accessed_storage_keys: {},
                },
            ],
            storage_read_values: [
                StarkFelt(
                    "0x041a78e741e5af2fec34b695679bc6891742439f7afb8484ecd7766661ad02cf",
                ),
            ],
            accessed_storage_keys: {
                StorageKey(
                    PatriciaKey(
                        StarkFelt(
                            "0x0238cf5ef6d6264a50d29a47fdf07ec9b7a8e9873214fa58179c5bb40933fdcb",
                        ),
                    ),
                ),
            },
        },
    ],
    storage_read_values: [],
    accessed_storage_keys: {},
}

```

File: cairo-contracts/src/test_contracts/emit_multiple_events_across_contracts.cairo (hunk 2 above)
Comment (m-kus) [reply cairo-contracts/src/test_contracts/emit_multiple_events_across_contracts.cairo:44]: Probably was lunch time ;)

Comment (tdelabro) [reply cairo-contracts/src/test_contracts/emit_multiple_events_across_contracts.cairo:44]: haha no that's cool !

**Review by tdelabro — APPROVED** [review]

**Comment by tdelabro** [comment]
I will merge as soon as the CI is fixed

**Pull request merged**